common/src/chain/transaction/transaction_index/mod.rs

```diff
@@ -144,40 +144,23 @@ pub struct TxMainChainIndex {
     spent: Vec<OutputSpentState>,
 }
 
-// TODO: This function should probably operate on the whole block at once.
-//  I.e. take a block in and return a sequence of transaction positions.
-//  This way, we have to ask for each transaction separately and every time
-//  the whole block is encoded, giving O(N^2) complexity in number of transactions
-//  which rather unpleasant. Also the implementation could be improved by
-//  only asking about offsets, leveraging Encode::encoded_size method, since
-//  we are only interested in offsets in the encoded stream, not the contents.
-pub fn calculate_tx_index_from_block(
-    block: &Block,
-    tx_num: usize,
-) -> Result<TxMainChainIndex, TxMainChainIndexError> {
-    let tx = block
-        .transactions()
-        .get(tx_num)
-        .ok_or_else(|| TxMainChainIndexError::InvalidTxNumberForBlock(tx_num, block.get_id()))?;
-    let enc_block = block.encode();
-    let enc_tx = tx.encode();
-    let offset_tx = enc_block
-        .windows(enc_tx.len())
-        .enumerate()
-        .find_map(|(window_num, enc_data)| (enc_data == enc_tx).then_some(window_num))
-        .ok_or_else(|| TxMainChainIndexError::SerializationInvariantError(block.get_id()))?
-        .try_into()
-        .expect("Number conversion from usize to u32 should not fail here (1)");
-
-    let tx_position = TxMainChainPosition::new(block.get_id(), offset_tx);
-
-    TxMainChainIndex::new(
-        SpendablePosition::from(tx_position),
-        tx.outputs()
-            .len()
-            .try_into()
-            .expect("Number conversion from usize to u32 should not fail here (3)"),
-    )
+pub fn calculate_tx_offsets_in_block(block: &Block) -> Vec<u32> {
+    // Please add new test if new block format version is added
```

Comment:
You don't need to add this comment. The match below will yell at whoever tries to ignore this ;-)

Reply:
The better way is to create a superficial match there that triggers an error :wink: